samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -29,97 +28,116 @@
    * {@link WindowPane}s.
    * <p>
    * Use the {@link org.apache.samza.operators.windows.Windows} helper methods to create the appropriate windows.
    *
-   * @param window the window to group and process {@link MessageEnvelope}s from this {@link MessageStream}
-   * @param <K> the type of key in the {@link MessageEnvelope} in this {@link MessageStream}. If a key is specified,
+   * @param window the window to group and process messages from this {@link MessageStream}
+   * @param <K> the type of key in the message in this {@link MessageStream}. If a key is specified,
    *            panes are emitted per-key.
    * @param <WV> the type of value in the {@link WindowPane} in the transformed {@link MessageStream}
-   * @param <WM> the type of {@link WindowPane} in the transformed {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
-  <K, WV, WM extends WindowPane<K, WV>> MessageStream<WM> window(Window<M, K, WV, WM> window);
+  <K, WV> MessageStream<WindowPane<K, WV>> window(Window<M, K, WV> window);
 
   /**
    * Joins this {@link MessageStream} with another {@link MessageStream} using the provided pairwise {@link JoinFunction}.
    * <p>
    * We currently only support 2-way joins.
    *
    * @param otherStream the other {@link MessageStream} to be joined with
-   * @param joinFn the function to join {@link MessageEnvelope}s from this and the other {@link MessageStream}
+   * @param joinFn the function to join messages from this and the other {@link MessageStream}
    * @param <K> the type of join key
-   * @param <OM> the type of {@link MessageEnvelope}s in the other stream
-   * @param <RM> the type of {@link MessageEnvelope}s resulting from the {@code joinFn}
+   * @param <OM> the type of messages in the other stream
+   * @param <RM> the type of messages resulting from the {@code joinFn}
    * @return the joined {@link MessageStream}
    */
-  <K, OM extends MessageEnvelope<K, ?>, RM extends MessageEnvelope> MessageStream<RM> join(MessageStream<OM> otherStream,
-      JoinFunction<M, OM, RM> joinFn);
+  <K, OM, RM> MessageStream<RM> join(MessageStream<OM> otherStream, JoinFunction<K, M, OM, RM> joinFn);
 
   /**
    * Merge all {@code otherStreams} with this {@link MessageStream}.
    * <p>
-   * The merging streams must have the same {@link MessageEnvelope} type {@code M}.
+   * The merging streams must have the same messages of type {@code M}.
    *
    * @param otherStreams  other {@link MessageStream}s to be merged with this {@link MessageStream}
    * @return  the merged {@link MessageStream}
    */
   MessageStream<M> merge(Collection<MessageStream<M>> otherStreams);
-  
+
+  /**
+   * Send the input message to an output {@link org.apache.samza.system.SystemStream} and consume it as input {@link MessageStream} again.
+   *
+   * Note: this is an transform function only used in logic DAG. In a physical DAG, this is either translated to a NOOP function, or a {@code MessageStream#sendThrough} function.
```

Comment:
I wonder if this is important to call out? Do we define anywhere what the logical and physical DAGs are?